pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,171 @@
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class ControllerLeaderLocatorIntegrationTest extends ControllerTest {
+  private static long TIMEOUT_IN_MS = 1000_000L;
+  private HashMap<Integer, String> _partitionToTableMap;
+
+  @BeforeClass
+  public void setUp() {
+    startZk();
+    startController();
+
+    ControllerLeaderLocator.create(_helixManager);
+
+    findTableNamesForAllPartitions();
+  }
+
+  @Test
+  public void testControllerLeaderLocator() {
+    String testTableName1 = "testTable";
+    String testTableName2 = "testTable2";
+    ControllerLeaderLocator controllerLeaderLocator = ControllerLeaderLocator.getInstance();
+
+    Pair<String, Integer> pair = controllerLeaderLocator.getControllerLeader(testTableName1);
+    // Before resource config is enabled, use helix leader
+    Assert.assertNotNull(pair);
+    Assert.assertEquals(pair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) pair.getSecond(), ControllerTest.DEFAULT_CONTROLLER_PORT);
+
+    // Enable lead controller resource
+    enableResourceConfigForLeadControllerResource(true);
+
+    // Mock the behavior that 40 seconds have passed.
+    controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // After resource config is enabled, use the lead controller in the resource
+    pair = controllerLeaderLocator.getControllerLeader(testTableName1);
+    Assert.assertNotNull(pair);
+    Assert.assertEquals(pair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) pair.getSecond(), ControllerTest.DEFAULT_CONTROLLER_PORT);
+
+    // Fetch the pair for test table 2, which should be the same as the 1st pair.
+    Pair<String, Integer> secondPair = controllerLeaderLocator.getControllerLeader(testTableName2);
+    Assert.assertNotNull(secondPair);
+    Assert.assertEquals(pair.getFirst(), secondPair.getFirst());
+    Assert.assertEquals(pair.getSecond(), secondPair.getSecond());
+
+    ControllerConf secondControllerConfig = getDefaultControllerConfiguration();
+    secondControllerConfig.setControllerPort(Integer.toString(DEFAULT_CONTROLLER_PORT + 1));
+    ControllerStarter secondControllerStarter = new ControllerStarter(secondControllerConfig);
+    secondControllerStarter.start();
+
+    TestUtils
+        .waitForCondition(aVoid -> secondControllerStarter.getHelixResourceManager().getHelixZkManager().isConnected(),
+            TIMEOUT_IN_MS, "Failed to start the second controller");
+
+    Set<String> resultSet = new HashSet<>();
+    TestUtils.waitForCondition(aVoid -> {
+      resultSet.clear();
+      for (Map.Entry<Integer, String> entry : _partitionToTableMap.entrySet()) {
+        // Mock the behavior that 40 seconds have passed.
+        controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
```

Comment:
you should not need this once you merge the other PR

Reply:
Rebased with latest master branch.